livox_ros2_driver/livox_ros2_driver/lddc.cpp

```diff
@@ -712,433 +702,386 @@ void Lddc::initializeDiagnostics()
+  }
+
   updater_.force_update();
 }
 
-void Lddc::checkTemperature(diagnostic_updater::DiagnosticStatusWrapper & stat)
+void Lddc::getLidarByBroadcastcode(std::pair<std::string, DeviceInfo *> &lidar,
+                                   const std::string &broadcast_code)
 {
-  if (lidar_count_ == 0) {
-    stat.summary(DiagStatus::WARN, "No LiDARs Connected");
-    return;
+  for (const auto & connected_lidar : lds_->connected_lidars_) {
+    const std::string connected_broadcast_code = connected_lidar.first;
+    if (connected_broadcast_code == broadcast_code) lidar = connected_lidar;
   }
+}
 
-  int whole_level = DiagStatus::OK;
-
-  for (const auto & lidar : lds_->connected_lidars_) {
-    int level = DiagStatus::OK;
+void Lddc::checkTemperature(diagnostic_updater::DiagnosticStatusWrapper &stat,
+                            const std::string & broadcast_code)
+{
+  int level = DiagStatus::OK;
+  std::string error_str = "OK";
 
-    const auto & broadcast_code = lidar.first;
-    const auto & device_info = lidar.second;
+  std::pair<std::string, DeviceInfo*> lidar;
+  getLidarByBroadcastcode(lidar, broadcast_code);
 
-    if (device_info == nullptr) {
-      stat.add(broadcast_code, "disconnected");
-      continue;
-    }
-
-    if (device_info->state == kLidarStateInit) {
-      stat.addf(broadcast_code, "%d%%", device_info->status.progress);
-      continue;
-    }
+  const auto & device_info = lidar.second;
 
+  if (device_info == nullptr) {
+    error_str = "disconnected";
+  } else if (device_info->state == kLidarStateInit) {
+    error_str = fmt::format("progress: %d%%", device_info->status.progress);
+  } else {
     TemperatureStatus status = static_cast<TemperatureStatus>(
       device_info->status.status_code.lidar_error_code.temp_status);
     if (status == TemperatureStatus::HighOrLow) {
       level = DiagStatus::WARN;
     } else if (status == TemperatureStatus::ExtremelyHighOrLow) {
       level = DiagStatus::ERROR;
     }
-
-    stat.add(broadcast_code, temperature_dict_.at(level));
-    whole_level = std::max(whole_level, level);
+    error_str = temperature_dict_.at(level);
   }
 
-  stat.summary(whole_level, temperature_dict_.at(whole_level));
+  stat.add(broadcast_code, error_str);
+  stat.summary(level, error_str);
 }
 
-void Lddc::checkVoltage(diagnostic_updater::DiagnosticStatusWrapper & stat)
+void Lddc::checkVoltage(diagnostic_updater::DiagnosticStatusWrapper &stat,
+                        const std::string &broadcast_code)
 {
-  if (lidar_count_ == 0) {
-    stat.summary(DiagStatus::WARN, "No LiDARs Connected");
-    return;
-  }
-
-  int whole_level = DiagStatus::OK;
-
-  for (const auto & lidar : lds_->connected_lidars_) {
-    int level = DiagStatus::OK;
+  int level = DiagStatus::OK;
+  std::string error_str = "OK";
 
-    const auto & broadcast_code = lidar.first;
-    const auto & device_info = lidar.second;
+  std::pair<std::string, DeviceInfo*> lidar;
+  getLidarByBroadcastcode(lidar, broadcast_code);
```

Comment:
With https://github.com/tier4/livox_ros2_driver/pull/14/files#r642506730, this would be like this.

```cpp
const auto device_info = getLidarByBroadcastcode(broadcast_code);

if (!device_info) {
    error_str = "disconnected";
} else if (device_info->state == kLidarStateInit) {
    error_str = fmt::format("progress: %d%%", device_info->status.progress);
}
```

Reply:
I had to use `*device_info)->state` when using `boost::optional`.